apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1342,21 +1377,265 @@ private void StartPush()
                     _pushInFlight += acks.Count;
                     _concurrentPushCount += 1;
                     Task ackTask = _client.AcknowledgeAsync(_subscriptionName, acks, _hardStopCts.Token);
-                    Add(ackTask, Next(false, () => HandleAckResponse(ackTask, acks, null, null)));
+                    Add(ackTask, Next(false, _exactlyOnceDeliveryEnabled ? () => HandleAckResponseForExactlyOnce(ackTask, acks, null, null) : () => HandleAckResponse(ackTask, acks, null, null)));
                 }
                 if (extends.Count > 0)
                 {
                     _pushInFlight += extends.Count;
                     _concurrentPushCount += 1;
                     Task extendTask = _client.ModifyAckDeadlineAsync(_subscriptionName, extends.Select(x => x.Id), _modifyDeadlineSeconds, _hardStopCts.Token);
-                    Add(extendTask, Next(false, () => HandleAckResponse(extendTask, null, null, extends)));
+                    Add(extendTask, Next(false, _exactlyOnceDeliveryEnabled ? () => HandleAckResponseForExactlyOnce(extendTask, null, null, extends) : () => HandleAckResponse(extendTask, null, null, extends)));
                 }
                 if (nacks.Count > 0)
                 {
                     _pushInFlight += nacks.Count;
                     _concurrentPushCount += 1;
                     Task nackTask = _client.ModifyAckDeadlineAsync(_subscriptionName, nacks, 0, _hardStopCts.Token);
-                    Add(nackTask, Next(false, () => HandleAckResponse(nackTask, null, nacks, null)));
+                    Add(nackTask, Next(false, _exactlyOnceDeliveryEnabled ? () => HandleAckResponseForExactlyOnce(nackTask, null, nacks, null) : () => HandleAckResponse(nackTask, null, nacks, null)));
+                }
+            }
+
+            private void Retry(IEnumerable<string> temporaryFailures, Action<IEnumerable<string>> actionToRetry, int durationInSeconds = 600)
+            {
+                // TemporaryFailures may be empty but will never be null, so null check is not needed.
+                // By design, only items in temporaryFailures that exist in _retyrableIds will be retried.
+                // If id is not in _retryableIds dictionary, it is either already succeeded or not retryable.
+                // Failure may be non-retryable due to exceeded time duration or that failure may have become permanent.
+                if (temporaryFailures.Any())
```

Comment:
I'd consider inverting the logic: if there aren't any temporary failures, return.
Likewise later on, if we've reached the deadline, return. That keeps nesting to a minimum, and makes the "we're done" cases really clear.